lib/srv/discovery/discovery.go

```diff
@@ -873,16 +858,14 @@ func (s *Server) heartbeatEICEInstance(instances *server.EC2Instances) {
 			continue
 		}
 
-		serverInfo, err := types.ServerInfoForServer(eiceNode)
-		if err != nil {
-			s.Log.WithField("instance_id", ec2Instance.InstanceID).Warnf("Error converting to Teleport ServerInfo: %v", err)
+		existingNode, err := s.nodeWatcher.GetNode(s.ctx, eiceNode.GetName())
+		if err != nil && !trace.IsNotFound(err) {
+			s.Log.Warnf("Error finding the existing node with name %q: %v", eiceNode.GetName(), err)
 			continue
 		}
 
-		if existingNode, found := existingEICEServersMap[serverInfo.GetName()]; found {
-			// Re-use the same Name so that `UpsertNode` replaces the node.
-			eiceNode.SetName(existingNode.GetName())
-
+		// EICE Node's Name are deterministic (based on the Account and Instance ID).
+		if existingNode != nil {
```

Comment:
Merge this with the subsequent conditional and do `if existingNode != nil && existingNode.Expiry().After(...) && ...`

Reply:
I actually prefer the current format, but that's fine :+1: